Update JSON serialization and deserialization code to use ``orjson`` library

Kami opened the pull request:
This pull request updates code base to use ``orjson`` library for serialization and de-serialization everywhere.

## Background, Context

Right now StackStorm code-base uses native python ``json`` library for serializing / de-serialiazing JSON. Only exception to that is our version of ``copy.deepcopy`` function which uses ``ujson`` since it's up to 10x faster than native ``copy.deepcopy``.

Now that we have dropped support for Python 2, we can look at using ``orjson`` (https://github.com/ijl/orjson) everywhere. ``orjson`` should be more or less backward compatible with ``json``, but also quite a bit faster (even faster than ``ujson`` and unlike ``ujson``, it's also actively maintainer).

## Proposed Change

I will update / add new ``json_encode()`` and ``json_decode()`` function which should be used everywhere in the code where we need to deal with JSON.

Those functions will utilize new feature flag / config option which will allow user to specify which json library to use. This feature flag should mostly serve as a safe-guard so user can revert back to native ``json`` library in case issues arise once the code is out.

## TODO

- [x] Add new ``json_encode`` and ``json_decode`` functions which handle json encoding and decoding using diffrent backends (json, ujson, orjson)
- [x] Add config option / feature flag which dictates which JSON library to use
- [x] Update API layer code (requests, responses) to use json_encode and decode
- [x] Update rest of the code to use json_encode and decode
- [x] Update ``fast_deepcopy`` to use orjson
- [x] Move ujson, simplejson from requirements to tests-requirements.txt
- [x] Decide if / when ``orjson`` should be default for the new config option value

blag commented:
A few thoughts:

- We should try to avoid custom magic methods, as they are subject to break without notice. See [this SO post](https://stackoverflow.com/a/9302851) and [the Python documentation itself](https://docs.python.org/3/reference/lexical_analysis.html#reserved-classes-of-identifiers)
- I don't like the idea of configuring a JSON parser backend, because this is something that is used so frequently within StackStorm that every future non-negligible bug report will need to post the config value for this, and users have a difficult enough time following our current directions for reporting issues - they aren't going to post their configs by default, so reported issues are going to have much more of a back-and-forth to get anywhere.
- That being said, the orjson library emphasizes "correctness" and does not emphasize "compatibility". JSON is so simple that I hope there isn't a lot of invalid-but-still-widely-parsable JSON out there, but if there is, then switching to orjson could expose JSON encoding issues with _other_ systems that get reported as regressions in StackStorm because all a user will see is that "it worked in the previous version of StackStorm". So if we switch to orjson, we need to be prepared to handle issues like that.

I think this is not a configuration option we should give to users. We should trust our tests and testing procedures to catch and uncover JSON parsing issues and switch to the backend that we (developers) think gives the most benefit to our users. I do think that orjson is probably the best library for this, and I'm happy to switch to it, but I want us to be prepared for any issues that crop up, and I don't think we should make this user-configurable.

Kami commented:
@blag 

> We should try to avoid custom magic methods, as they are subject to break without notice. See this SO post and the Python documentation itself

Which magic method are you referring to? If you mean ``__json__``, that's an existing behavior and nothing I'm changing (and also nothing I want to change at this point to keep the scope down).

> That being said, the orjson library emphasizes "correctness" and does not emphasize "compatibility". JSON is so simple that I hope there isn't a lot of invalid-but-still-widely-parsable JSON out there, but if there is, then switching to orjson could expose JSON encoding issues with other systems that get reported as regressions in StackStorm because all a user will see is that "it worked in the previous version of StackStorm". So if we switch to orjson, we need to be prepared to handle issues like that.

That's exactly the reason why that configuration option / feature flag is there.

In general, as mentioned in the discussions thread, this change already exposed some issues with incorrect code in some places which mostly results from incorrect and inconsistent conversion between bytes <-> unicode.

Incorrect in a sense that right now the code doesn't throw, but the response will be "incorrect" - incorrect in a sense that some string data in responses will contain ``b""`` prefix which it shouldn't.

Kami commented:
And in general I also think ``orjson`` is the best choice from the strictness perspective.

In the past, we have been bit so many times by non-strict nature of various casting and other code. Problem with using best effort, non strict, etc. approach is that it may work for a lot of scenarios, but when it breaks, it's very hard to troubleshoot and fix it.

Kami commented:
I believe packages step is failing because we are using older pip version which doesn't include support for manywheel wheel format orjson uses so it tries to compile it from scratch instead of using a wheel.

Will look into fixing that.

Kami commented:
Related PR to get st2-packages build to pass - https://github.com/StackStorm/st2packaging-dockerfiles/pull/103.

blag commented:
PR #5123 updates pip to 20.3.x, which should handle manylinux wheels:

| `manylinux` tag | Client-side pip version required |
| ----------------- | -------------------------------- |
| `manylinux_x_y` | **pip >= 20.3**                      |
| `manylinux2014` | pip >= 19.3                      |
| `manylinux2010` | pip >= 19.0                      |
| `manylinux1`    | pip >= 8.1.0                     |

Kami commented:
@blag In this case the issue are build containers not StackStorm/st2, since pip 20.x is sufficient for orjson manywheel2014 (https://github.com/StackStorm/st2packaging-dockerfiles/pull/103#issuecomment-778836022, https://github.com/StackStorm/st2packaging-dockerfiles/pull/103#issuecomment-778844161).

For now, I can wait on https://github.com/StackStorm/st2packaging-dockerfiles/pull/98, but if things drag to long, another option is to just upgrade to v19.3 in the mean time in those Dockerfiles.

Kami commented:
For completeness sake, here are also microbenchmarks results for ``fast_deepcopy`` - basically something we were already aware of and that's why that change has been made originally, but now we also have micro benchmarks in place.

![Screenshot from 2021-02-14 22-27-28](https://user-images.githubusercontent.com/125088/107889583-f75c1180-6f13-11eb-832c-97f2050d2089.png)

As you can see, in this case ``orjson`` doesn't bring all that much of an improvement on top of ``ujson``, but any improvement we get on top of ``ujson`` is still good "free" so we should take it :)

Once I get access to some large executions (executions with large result field) in JSON format, I will also hook it up to that micro benchmarks (since that will represent some more real-life like data).

Kami commented:
I've added a changelog entry and a note on backward compatibility.

The change should be fully backward compatible, only difference is that API responses now would be pretty formatted (indented) by default.

The reason for that is that pretty formatting / indenting adds some overhead and it can be done on the client side, if needed.

If people feel stronlgy about it, it can be enabled (or perhaps we can add config option for it / enable it when ``debug`` flag is set to true).

Kami commented:
I also updated Python runner to use ``orjson`` for serializing results - this results in faster execution time for Python runner actions which return large results (as described in the comment, we don't want to use ``json_encode`` there since this would add dependency on config option which is what we want to avoid in the wrapper process since it adds overhead to import time and making sure that overhead is as low as possible there is essential).

To put things into perspective - Python runner running an action which returns 5 MB json + parsing the result (which doesn't include saving it to the database) takes around 500 ms with current master and around 300 with this PR.

cognifloyd commented on st2common/benchmarks/micro/test_fast_deepcopy.py:
```diff
@@ -0,0 +1,152 @@
+# limitations under the License.
+
+"""
+Micro benchmark which compares the performance of or fast_deepcopy implementation using different
```
```suggestion
Micro benchmark which compares the performance of our fast_deepcopy implementation using different
```

cognifloyd reviewed (approved):
:tada:

I noticed one minor docstring typo, but whatever. This looks awesome.

cognifloyd reviewed (commented):
You can drop the changes in tools/config_gen.py now that `json_library` is a module constant instead of a config option.

Kami commented:
@blag Can I please get a review when you get a chance.

I removed the config option, changed it to a module level constant so we can still exercise it in compatibility tests.

blag commented:
On mobile right now, will review later today. 👍

Kami commented:
@blag Thanks.

I'm somewhat confused by u16 e2e test failure. Is this a known issue, or a race or smth?

Looking at the test output, I'm quite confused - https://gist.github.com/Kami/0d83da44842daa3cbb9dcef670299df6. That alias doesn't have representation (https://github.com/StackStorm-Exchange/stackstorm-st2/blob/master/aliases/actions_list.yaml) so that assertion looks wrong to me. But how did it pass before?

And my code touched none of that.

I would assume it was race or smth, but assertion doesn't seem to make sense since it doesn't add up with the actual alias defined in stackstorm-st2 pack.

To me it seems like that bats test should check for ``st2 list {{ limit=10 }} actions - List available StackStorm actions.``, but I could be missing something (aka we are trying to assert on the actual hubot response and not the start up string where it lists the loaded commands).

Kami commented:
OK, ignore my comment above - I'm still digging into end to end tests output and the instance itself...

Kami commented:
OK, yeah I was able to reproduce the issue - it's indeed locale related on Ubuntu 16.04.

If it's set to utf8 it works fine, but it's set to ascii it blows up due to change in how orjson serialized unicodes (it doesn't require serialized it to ascii escape sequences, but utilizes unicode directly which is preferred).

I'm having troubles dynamically setting ``PYTHONIOENCODING=utf8`` for action runner wrapper inside the Python runner. It doesn't play nicely with the eventlet patched subprocess module, but I'm still looking into other possible solutions.

Kami commented:
OK, finally the whole build is green :tada: :champagne:

cognifloyd commented on tools/config_gen.py:
```diff
@@ -41,6 +41,7 @@
 ]
 
 SKIP_GROUPS = ["api_pecan", "rbac", "results_tracker"]
+SKIP_OPTIONS = ["json_library"]
```
Now that `json_library` is not an option, do you still want to leave this here?

m4dcoder replied:
I agree. We talked about not having user config options for json lib.

Kami replied:
Yeah, the config was removed but the this flag was left here so we can also support ignoring specific config options (previously this script didn't support that).

I will change ``SKIP_OPTIONS`` to an empty list.

m4dcoder commented on contrib/runners/action_chain_runner/action_chain_runner/action_chain_runner.py:
```diff
@@ -48,7 +48,7 @@
 from st2common.util import param as param_utils
 from st2common.util.config_loader import get_config
-from st2common.util.ujson import fast_deepcopy
+from st2common.util.deep_copy import fast_deepcopy
```
Can we call this module json util or something so not to be confused with copy.deepcopy?

Kami replied:
I intentionally renamed it to ``deep_copy`` since I think previous name was a bad one (I know, I picked it originally :D) - previous name was leaking implementation details which should not matter to the end user.

I think ``deep_copy`` is a better name since it conveys what the module is used for - deep copying dictionaries.

And the module is not JSON specific either, it's can deep copy arbitrary dictionaries with simple types. ``orjson`` is just an implementation detail of that function.

m4dcoder replied:
I don't want this to be confused with copy.deepcopy which is not json or dict specific.

Kami replied:
Maybe I can rename it to ``fast_deepcopy_dict`` then?

Kami replied:
And I guess I need to be more explicitly - it doesn't just support dicts either (that's just how we use it).

It supports any kind of value as long as it only contains simple types (so no class instances, etc.).

Kami replied:
Per discussion on Slack, I pushed a couple of changes - add some more tests, update function docstring comment, rename it to ``fast_deepcopy_dict``.

As discussed, it can also be used on other simple values (think lists) and not just dictionaries, but using it on dictionaries with simple value types is our primary use case for it so I think that name is an OK compromise for now.

m4dcoder commented on contrib/runners/python_runner/python_runner/python_action_wrapper.py:
```diff
@@ -28,6 +28,12 @@
+# on config option which we don't parse for the action wrapper since it speeds things down - action
+# wrapper should rely on as little imports as possible to make Python runner executions fast -
+# that's very important.
+import orjson
```
Could we use jsonify json encode/decode methods here instead of referencing orjson directly here?

Kami replied:
EDIT: Ignore my comment above, I thought I was looking at another PR.

Actually, no, we can't do that because of the comment above the import line.

> 
> NOTE: We intentionally use orjson directly here instead of json_encode - orjson.dumps relies
> on config option which we don't parse for the action wrapper since it speeds things down - action
> wrapper should rely on as little imports as possible to make Python runner executions fast -
> that's very important.

In short, in python action wrapper every 10s of milliseconds make a difference and importing that module adds a bunch of overhead since it has import time side affects due to loading oslo config, etc.

m4dcoder replied:
Ack.

m4dcoder reviewed (changes requested):
Mostly OK but some clean up requests.

m4dcoder reviewed (approved):
@Kami Thank you! This is going to significantly improve st2.
